In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib as mpl


##### we taken dataset from Quikr

In [2]:
car=pd.read_csv("E:/car_prediction project/car_Raw_dataset.csv")

In [3]:
car

name    company  year          Price  \
0      Hyundai Santro Xing XO eRLX Euro III    Hyundai  2007         80,000   
1                   Mahindra Jeep CL550 MDI   Mahindra  2006       4,25,000   
2                Maruti Suzuki Alto 800 Vxi     Maruti  2018  Ask For Price   
3    Hyundai Grand i10 Magna 1.2 Kappa VTVT    Hyundai  2014       3,25,000   
4          Ford EcoSport Titanium 1.5L TDCi       Ford  2014       5,75,000   
..                                      ...        ...   ...            ...   
887                                      Ta       Tara  zest       3,10,000   
888                     Tata Zest XM Diesel       Tata  2018       2,60,000   
889                      Mahindra Quanto C8   Mahindra  2013       3,90,000   
890                Honda Amaze 1.2 E i VTEC      Honda  2014       1,80,000   
891               Chevrolet Sail 1.2 LT ABS  Chevrolet  2014       1,60,000   

     kms_driven fuel_type  
0    45,000 kms    Petrol  
1        40 kms    Diesel  
2    22,000 kms    Petrol  
3    28,000 kms    Petrol  
4    36,000 kms    Diesel  
..          ...       ...  
887         NaN       NaN  
888  27,000 kms    Diesel  
889  40,000 kms    Diesel  
890      Petrol       NaN  
891      Petrol       NaN  

[892 rows x 6 columns]

In [4]:
car.shape

(892, 6)

In [5]:
car.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 892 entries, 0 to 891
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   name        892 non-null    object
 1   company     892 non-null    object
 2   year        892 non-null    object
 3   Price       892 non-null    object
 4   kms_driven  840 non-null    object
 5   fuel_type   837 non-null    object
dtypes: object(6)
memory usage: 41.9+ KB


In [6]:
car.describe()

name company  year          Price  kms_driven fuel_type
count          892     892   892            892         840       837
unique         525      48    61            274         258         3
top     Honda City  Maruti  2015  Ask For Price  45,000 kms    Petrol
freq            13     235   117             35          30       440

In [7]:
car['year'].unique()

array(['2007', '2006', '2018', '2014', '2015', '2012', '2013', '2016',
       '2010', '2017', '2008', '2011', '2019', '2009', '2005', '2000',
       '...', '150k', 'TOUR', '2003', 'r 15', '2004', 'Zest', '/-Rs',
       'sale', '1995', 'ara)', '2002', 'SELL', '2001', 'tion', 'odel',
       '2 bs', 'arry', 'Eon', 'o...', 'ture', 'emi', 'car', 'able', 'no.',
       'd...', 'SALE', 'digo', 'sell', 'd Ex', 'n...', 'e...', 'D...',
       ', Ac', 'go .', 'k...', 'o c4', 'zire', 'cent', 'Sumo', 'cab',
       't xe', 'EV2', 'r...', 'zest'], dtype=object)

In [8]:
car['Price'].unique()

array(['80,000', '4,25,000', 'Ask For Price', '3,25,000', '5,75,000',
       '1,75,000', '1,90,000', '8,30,000', '2,50,000', '1,82,000',
       '3,15,000', '4,15,000', '3,20,000', '10,00,000', '5,00,000',
       '3,50,000', '1,60,000', '3,10,000', '75,000', '1,00,000',
       '2,90,000', '95,000', '1,80,000', '3,85,000', '1,05,000',
       '6,50,000', '6,89,999', '4,48,000', '5,49,000', '5,01,000',
       '4,89,999', '2,80,000', '3,49,999', '2,84,999', '3,45,000',
       '4,99,999', '2,35,000', '2,49,999', '14,75,000', '3,95,000',
       '2,20,000', '1,70,000', '85,000', '2,00,000', '5,70,000',
       '1,10,000', '4,48,999', '18,91,111', '1,59,500', '3,44,999',
       '4,49,999', '8,65,000', '6,99,000', '3,75,000', '2,24,999',
       '12,00,000', '1,95,000', '3,51,000', '2,40,000', '90,000',
       '1,55,000', '6,00,000', '1,89,500', '2,10,000', '3,90,000',
       '1,35,000', '16,00,000', '7,01,000', '2,65,000', '5,25,000',
       '3,72,000', '6,35,000', '5,50,000', '4,85,000', '3,29,5

In [9]:
car['kms_driven'].unique()

array(['45,000 kms', '40 kms', '22,000 kms', '28,000 kms', '36,000 kms',
       '59,000 kms', '41,000 kms', '25,000 kms', '24,530 kms',
       '60,000 kms', '30,000 kms', '32,000 kms', '48,660 kms',
       '4,000 kms', '16,934 kms', '43,000 kms', '35,550 kms',
       '39,522 kms', '39,000 kms', '55,000 kms', '72,000 kms',
       '15,975 kms', '70,000 kms', '23,452 kms', '35,522 kms',
       '48,508 kms', '15,487 kms', '82,000 kms', '20,000 kms',
       '68,000 kms', '38,000 kms', '27,000 kms', '33,000 kms',
       '46,000 kms', '16,000 kms', '47,000 kms', '35,000 kms',
       '30,874 kms', '15,000 kms', '29,685 kms', '1,30,000 kms',
       '19,000 kms', nan, '54,000 kms', '13,000 kms', '38,200 kms',
       '50,000 kms', '13,500 kms', '3,600 kms', '45,863 kms',
       '60,500 kms', '12,500 kms', '18,000 kms', '13,349 kms',
       '29,000 kms', '44,000 kms', '42,000 kms', '14,000 kms',
       '49,000 kms', '36,200 kms', '51,000 kms', '1,04,000 kms',
       '33,333 kms', '33,600 kms', '5,

In [10]:
car['fuel_type'].unique()

array(['Petrol', 'Diesel', nan, 'LPG'], dtype=object)

In [11]:
car['name'].unique()

array(['Hyundai Santro Xing XO eRLX Euro III', 'Mahindra Jeep CL550 MDI',
       'Maruti Suzuki Alto 800 Vxi',
       'Hyundai Grand i10 Magna 1.2 Kappa VTVT',
       'Ford EcoSport Titanium 1.5L TDCi', 'Ford Figo', 'Hyundai Eon',
       'Ford EcoSport Ambiente 1.5L TDCi',
       'Maruti Suzuki Alto K10 VXi AMT', 'Skoda Fabia Classic 1.2 MPI',
       'Maruti Suzuki Stingray VXi', 'Hyundai Elite i20 Magna 1.2',
       'Mahindra Scorpio SLE BS IV', 'Audi A8', 'Audi Q7',
       'Mahindra Scorpio S10', 'Maruti Suzuki Alto 800',
       'Hyundai i20 Sportz 1.2', 'Maruti Suzuki Alto 800 Lx',
       'Maruti Suzuki Vitara Brezza ZDi', 'Maruti Suzuki Alto LX',
       'Mahindra Bolero DI', 'Maruti Suzuki Swift Dzire ZDi',
       'Mahindra Scorpio S10 4WD', 'Maruti Suzuki Swift Vdi BSIII',
       'Maruti Suzuki Wagon R VXi BS III',
       'Maruti Suzuki Wagon R VXi Minor',
       'Toyota Innova 2.0 G 8 STR BS IV', 'Renault Lodgy 85 PS RXL',
       'Skoda Yeti Ambition 2.0 TDI CR 4x2',
       'Maru

## Quality

- names are pretty inconsistent
- names have company names attached to it
- some names are spam like 'Maruti Ertiga showroom condition with' and 'Well mentained Tata Sumo'
- company: many of the names are not of any company like 'Used', 'URJENT', and so on.
- year has many non-year values
- year is in object. Change to integer
- Price has 'Ask for Price'
- Price has commas in its prices and is in object
- kms_driven has object values with kms at last.
- It has nan values and two rows have 'Petrol' in them
- fuel_type has nan values

## DATA BACKUP 

In [12]:
backup=car.copy()

In [13]:
car['year']

0      2007
1      2006
2      2018
3      2014
4      2014
       ... 
887    zest
888    2018
889    2013
890    2014
891    2014
Name: year, Length: 892, dtype: object

* year has many non-year values

In [14]:
car['year'].str.isnumeric()

0       True
1       True
2       True
3       True
4       True
       ...  
887    False
888     True
889     True
890     True
891     True
Name: year, Length: 892, dtype: bool

In [15]:
car=car[car['year'].str.isnumeric()]
car

name    company  year          Price  \
0      Hyundai Santro Xing XO eRLX Euro III    Hyundai  2007         80,000   
1                   Mahindra Jeep CL550 MDI   Mahindra  2006       4,25,000   
2                Maruti Suzuki Alto 800 Vxi     Maruti  2018  Ask For Price   
3    Hyundai Grand i10 Magna 1.2 Kappa VTVT    Hyundai  2014       3,25,000   
4          Ford EcoSport Titanium 1.5L TDCi       Ford  2014       5,75,000   
..                                      ...        ...   ...            ...   
886                    Toyota Corolla Altis     Toyota  2009       3,00,000   
888                     Tata Zest XM Diesel       Tata  2018       2,60,000   
889                      Mahindra Quanto C8   Mahindra  2013       3,90,000   
890                Honda Amaze 1.2 E i VTEC      Honda  2014       1,80,000   
891               Chevrolet Sail 1.2 LT ABS  Chevrolet  2014       1,60,000   

       kms_driven fuel_type  
0      45,000 kms    Petrol  
1          40 kms    Diesel  
2      22,000 kms    Petrol  
3      28,000 kms    Petrol  
4      36,000 kms    Diesel  
..            ...       ...  
886  1,32,000 kms    Petrol  
888    27,000 kms    Diesel  
889    40,000 kms    Diesel  
890        Petrol       NaN  
891        Petrol       NaN  

[842 rows x 6 columns]

# now change datatype int

In [16]:
car['year'].astype(int)

0      2007
1      2006
2      2018
3      2014
4      2014
       ... 
886    2009
888    2018
889    2013
890    2014
891    2014
Name: year, Length: 842, dtype: int32

In [17]:
car.info()

<class 'pandas.core.frame.DataFrame'>
Index: 842 entries, 0 to 891
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   name        842 non-null    object
 1   company     842 non-null    object
 2   year        842 non-null    object
 3   Price       842 non-null    object
 4   kms_driven  840 non-null    object
 5   fuel_type   837 non-null    object
dtypes: object(6)
memory usage: 46.0+ KB


In [18]:
car['year']=car['year'].astype(int)

C:\Users\ADMIN\AppData\Local\Temp\ipykernel_13516\152728128.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  car['year']=car['year'].astype(int)


In [19]:
car.info()

<class 'pandas.core.frame.DataFrame'>
Index: 842 entries, 0 to 891
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   name        842 non-null    object
 1   company     842 non-null    object
 2   year        842 non-null    int32 
 3   Price       842 non-null    object
 4   kms_driven  840 non-null    object
 5   fuel_type   837 non-null    object
dtypes: int32(1), object(5)
memory usage: 42.8+ KB


In [20]:
car=car[car['Price'] != "Ask For Price"]
car

name    company  year     Price  \
0      Hyundai Santro Xing XO eRLX Euro III    Hyundai  2007    80,000   
1                   Mahindra Jeep CL550 MDI   Mahindra  2006  4,25,000   
3    Hyundai Grand i10 Magna 1.2 Kappa VTVT    Hyundai  2014  3,25,000   
4          Ford EcoSport Titanium 1.5L TDCi       Ford  2014  5,75,000   
6                                 Ford Figo       Ford  2012  1,75,000   
..                                      ...        ...   ...       ...   
886                    Toyota Corolla Altis     Toyota  2009  3,00,000   
888                     Tata Zest XM Diesel       Tata  2018  2,60,000   
889                      Mahindra Quanto C8   Mahindra  2013  3,90,000   
890                Honda Amaze 1.2 E i VTEC      Honda  2014  1,80,000   
891               Chevrolet Sail 1.2 LT ABS  Chevrolet  2014  1,60,000   

       kms_driven fuel_type  
0      45,000 kms    Petrol  
1          40 kms    Diesel  
3      28,000 kms    Petrol  
4      36,000 kms    Diesel  
6      41,000 kms    Diesel  
..            ...       ...  
886  1,32,000 kms    Petrol  
888    27,000 kms    Diesel  
889    40,000 kms    Diesel  
890        Petrol       NaN  
891        Petrol       NaN  

[819 rows x 6 columns]

*  we removed those rows having "ASK FOR Price"

In [21]:
car['Price']

0        80,000
1      4,25,000
3      3,25,000
4      5,75,000
6      1,75,000
         ...   
886    3,00,000
888    2,60,000
889    3,90,000
890    1,80,000
891    1,60,000
Name: Price, Length: 819, dtype: object

NOw we want to remove commas in dataframe

In [22]:
car['Price']=car['Price'].str.replace(',','')

C:\Users\ADMIN\AppData\Local\Temp\ipykernel_13516\2888702569.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  car['Price']=car['Price'].str.replace(',','')


* now change it to int


In [23]:
car['Price']=car['Price'].astype(int)

C:\Users\ADMIN\AppData\Local\Temp\ipykernel_13516\1635868317.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  car['Price']=car['Price'].astype(int)


In [24]:
car.info()

<class 'pandas.core.frame.DataFrame'>
Index: 819 entries, 0 to 891
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   name        819 non-null    object
 1   company     819 non-null    object
 2   year        819 non-null    int32 
 3   Price       819 non-null    int32 
 4   kms_driven  819 non-null    object
 5   fuel_type   816 non-null    object
dtypes: int32(2), object(4)
memory usage: 38.4+ KB


In [25]:
car['kms_driven']

0        45,000 kms
1            40 kms
3        28,000 kms
4        36,000 kms
6        41,000 kms
           ...     
886    1,32,000 kms
888      27,000 kms
889      40,000 kms
890          Petrol
891          Petrol
Name: kms_driven, Length: 819, dtype: object

In [26]:
car['kms_driven'].str.split(' ')

0        [45,000, kms]
1            [40, kms]
3        [28,000, kms]
4        [36,000, kms]
6        [41,000, kms]
            ...       
886    [1,32,000, kms]
888      [27,000, kms]
889      [40,000, kms]
890           [Petrol]
891           [Petrol]
Name: kms_driven, Length: 819, dtype: object

In [27]:
car['kms_driven'].str.split(' ').str.get(0)

0        45,000
1            40
3        28,000
4        36,000
6        41,000
         ...   
886    1,32,000
888      27,000
889      40,000
890      Petrol
891      Petrol
Name: kms_driven, Length: 819, dtype: object

In [28]:
car['kms_driven']=car['kms_driven'].str.split(' ').str.get(0).str.replace(',','')

C:\Users\ADMIN\AppData\Local\Temp\ipykernel_13516\995198746.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  car['kms_driven']=car['kms_driven'].str.split(' ').str.get(0).str.replace(',','')


* now ther is some cateogorical data also 'Petrol'

In [29]:
car=car[car['kms_driven'].str.isnumeric()]
car

name   company  year   Price  \
0      Hyundai Santro Xing XO eRLX Euro III   Hyundai  2007   80000   
1                   Mahindra Jeep CL550 MDI  Mahindra  2006  425000   
3    Hyundai Grand i10 Magna 1.2 Kappa VTVT   Hyundai  2014  325000   
4          Ford EcoSport Titanium 1.5L TDCi      Ford  2014  575000   
6                                 Ford Figo      Ford  2012  175000   
..                                      ...       ...   ...     ...   
883              Maruti Suzuki Ritz VXI ABS    Maruti  2011  270000   
885               Tata Indica V2 DLE BS III      Tata  2009  110000   
886                    Toyota Corolla Altis    Toyota  2009  300000   
888                     Tata Zest XM Diesel      Tata  2018  260000   
889                      Mahindra Quanto C8  Mahindra  2013  390000   

    kms_driven fuel_type  
0        45000    Petrol  
1           40    Diesel  
3        28000    Petrol  
4        36000    Diesel  
6        41000    Diesel  
..         ...       ...  
883      50000    Petrol  
885      30000    Diesel  
886     132000    Petrol  
888      27000    Diesel  
889      40000    Diesel  

[817 rows x 6 columns]

In [30]:
car['kms_driven']=car['kms_driven'].astype(int)

C:\Users\ADMIN\AppData\Local\Temp\ipykernel_13516\1713770885.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  car['kms_driven']=car['kms_driven'].astype(int)


In [31]:
car.info()

<class 'pandas.core.frame.DataFrame'>
Index: 817 entries, 0 to 889
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   name        817 non-null    object
 1   company     817 non-null    object
 2   year        817 non-null    int32 
 3   Price       817 non-null    int32 
 4   kms_driven  817 non-null    int32 
 5   fuel_type   816 non-null    object
dtypes: int32(3), object(3)
memory usage: 35.1+ KB


In [32]:
car[car['fuel_type'].isna()]

name company  year   Price  kms_driven fuel_type
132  Toyota Corolla  Toyota  2009  275000       26000       NaN

In [33]:
car=car[~car['fuel_type'].isna()]

In [34]:
car

name   company  year   Price  \
0      Hyundai Santro Xing XO eRLX Euro III   Hyundai  2007   80000   
1                   Mahindra Jeep CL550 MDI  Mahindra  2006  425000   
3    Hyundai Grand i10 Magna 1.2 Kappa VTVT   Hyundai  2014  325000   
4          Ford EcoSport Titanium 1.5L TDCi      Ford  2014  575000   
6                                 Ford Figo      Ford  2012  175000   
..                                      ...       ...   ...     ...   
883              Maruti Suzuki Ritz VXI ABS    Maruti  2011  270000   
885               Tata Indica V2 DLE BS III      Tata  2009  110000   
886                    Toyota Corolla Altis    Toyota  2009  300000   
888                     Tata Zest XM Diesel      Tata  2018  260000   
889                      Mahindra Quanto C8  Mahindra  2013  390000   

     kms_driven fuel_type  
0         45000    Petrol  
1            40    Diesel  
3         28000    Petrol  
4         36000    Diesel  
6         41000    Diesel  
..          ...       ...  
883       50000    Petrol  
885       30000    Diesel  
886      132000    Petrol  
888       27000    Diesel  
889       40000    Diesel  

[816 rows x 6 columns]

In [35]:
car['name'].str.split(' ')

0        [Hyundai, Santro, Xing, XO, eRLX, Euro, III]
1                        [Mahindra, Jeep, CL550, MDI]
3      [Hyundai, Grand, i10, Magna, 1.2, Kappa, VTVT]
4              [Ford, EcoSport, Titanium, 1.5L, TDCi]
6                                        [Ford, Figo]
                            ...                      
883                  [Maruti, Suzuki, Ritz, VXI, ABS]
885                  [Tata, Indica, V2, DLE, BS, III]
886                          [Toyota, Corolla, Altis]
888                          [Tata, Zest, XM, Diesel]
889                            [Mahindra, Quanto, C8]
Name: name, Length: 816, dtype: object

In [36]:
car['name'].str.split(' ').str.slice(0,3)

0         [Hyundai, Santro, Xing]
1         [Mahindra, Jeep, CL550]
3           [Hyundai, Grand, i10]
4      [Ford, EcoSport, Titanium]
6                    [Ford, Figo]
                  ...            
883        [Maruti, Suzuki, Ritz]
885            [Tata, Indica, V2]
886      [Toyota, Corolla, Altis]
888              [Tata, Zest, XM]
889        [Mahindra, Quanto, C8]
Name: name, Length: 816, dtype: object

In [37]:
car['name']=car['name'].str.split(' ').str.slice(0,3).str.join(' ')

C:\Users\ADMIN\AppData\Local\Temp\ipykernel_13516\3677590539.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  car['name']=car['name'].str.split(' ').str.slice(0,3).str.join(' ')


In [38]:
car


name   company  year   Price  kms_driven fuel_type
0       Hyundai Santro Xing   Hyundai  2007   80000       45000    Petrol
1       Mahindra Jeep CL550  Mahindra  2006  425000          40    Diesel
3         Hyundai Grand i10   Hyundai  2014  325000       28000    Petrol
4    Ford EcoSport Titanium      Ford  2014  575000       36000    Diesel
6                 Ford Figo      Ford  2012  175000       41000    Diesel
..                      ...       ...   ...     ...         ...       ...
883      Maruti Suzuki Ritz    Maruti  2011  270000       50000    Petrol
885          Tata Indica V2      Tata  2009  110000       30000    Diesel
886    Toyota Corolla Altis    Toyota  2009  300000      132000    Petrol
888            Tata Zest XM      Tata  2018  260000       27000    Diesel
889      Mahindra Quanto C8  Mahindra  2013  390000       40000    Diesel

[816 rows x 6 columns]

In [39]:
car.info()

<class 'pandas.core.frame.DataFrame'>
Index: 816 entries, 0 to 889
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   name        816 non-null    object
 1   company     816 non-null    object
 2   year        816 non-null    int32 
 3   Price       816 non-null    int32 
 4   kms_driven  816 non-null    int32 
 5   fuel_type   816 non-null    object
dtypes: int32(3), object(3)
memory usage: 35.1+ KB


In [40]:
car.describe()

year         Price     kms_driven
count   816.000000  8.160000e+02     816.000000
mean   2012.444853  4.117176e+05   46275.531863
std       4.002992  4.751844e+05   34297.428044
min    1995.000000  3.000000e+04       0.000000
25%    2010.000000  1.750000e+05   27000.000000
50%    2013.000000  2.999990e+05   41000.000000
75%    2015.000000  4.912500e+05   56818.500000
max    2019.000000  8.500003e+06  400000.000000

* we can se our 75% car is less than 5lakh and maximum car price is 85 lakh so , definate ther is **outlier**.

In [41]:
car[car['Price']>5e6]

name   company  year    Price  kms_driven fuel_type
562  Mahindra XUV500 W6  Mahindra  2014  8500003       45000    Diesel

### this is the outlier , only one car is that above 50lakh price, so remove that outlier

In [42]:
car=car[car['Price']<5e6]
car

name   company  year   Price  kms_driven fuel_type
0       Hyundai Santro Xing   Hyundai  2007   80000       45000    Petrol
1       Mahindra Jeep CL550  Mahindra  2006  425000          40    Diesel
3         Hyundai Grand i10   Hyundai  2014  325000       28000    Petrol
4    Ford EcoSport Titanium      Ford  2014  575000       36000    Diesel
6                 Ford Figo      Ford  2012  175000       41000    Diesel
..                      ...       ...   ...     ...         ...       ...
883      Maruti Suzuki Ritz    Maruti  2011  270000       50000    Petrol
885          Tata Indica V2      Tata  2009  110000       30000    Diesel
886    Toyota Corolla Altis    Toyota  2009  300000      132000    Petrol
888            Tata Zest XM      Tata  2018  260000       27000    Diesel
889      Mahindra Quanto C8  Mahindra  2013  390000       40000    Diesel

[815 rows x 6 columns]

* now we can see in our data index is not  proper 
2,5,884,887 rows are droped , so we need to reset index 

In [43]:
car.reset_index()

index                    name   company  year   Price  kms_driven  \
0        0     Hyundai Santro Xing   Hyundai  2007   80000       45000   
1        1     Mahindra Jeep CL550  Mahindra  2006  425000          40   
2        3       Hyundai Grand i10   Hyundai  2014  325000       28000   
3        4  Ford EcoSport Titanium      Ford  2014  575000       36000   
4        6               Ford Figo      Ford  2012  175000       41000   
..     ...                     ...       ...   ...     ...         ...   
810    883      Maruti Suzuki Ritz    Maruti  2011  270000       50000   
811    885          Tata Indica V2      Tata  2009  110000       30000   
812    886    Toyota Corolla Altis    Toyota  2009  300000      132000   
813    888            Tata Zest XM      Tata  2018  260000       27000   
814    889      Mahindra Quanto C8  Mahindra  2013  390000       40000   

    fuel_type  
0      Petrol  
1      Diesel  
2      Petrol  
3      Diesel  
4      Diesel  
..        ...  
810    Petrol  
811    Diesel  
812    Petrol  
813    Diesel  
814    Diesel  

[815 rows x 7 columns]

* previus index is also there so, to drop that we use (drop=True)

In [44]:
car=car.reset_index(drop=True)

## Now the above Data is clean
* save this data

In [45]:
car.to_csv("cleaned car.csv")

# MODEL

In [46]:
x=car.drop(columns='Price')
# Y is our Target column
y=car['Price']

In [47]:
x

name   company  year  kms_driven fuel_type
0       Hyundai Santro Xing   Hyundai  2007       45000    Petrol
1       Mahindra Jeep CL550  Mahindra  2006          40    Diesel
2         Hyundai Grand i10   Hyundai  2014       28000    Petrol
3    Ford EcoSport Titanium      Ford  2014       36000    Diesel
4                 Ford Figo      Ford  2012       41000    Diesel
..                      ...       ...   ...         ...       ...
810      Maruti Suzuki Ritz    Maruti  2011       50000    Petrol
811          Tata Indica V2      Tata  2009       30000    Diesel
812    Toyota Corolla Altis    Toyota  2009      132000    Petrol
813            Tata Zest XM      Tata  2018       27000    Diesel
814      Mahindra Quanto C8  Mahindra  2013       40000    Diesel

[815 rows x 5 columns]

In [48]:
y

0       80000
1      425000
2      325000
3      575000
4      175000
        ...  
810    270000
811    110000
812    300000
813    260000
814    390000
Name: Price, Length: 815, dtype: int32

In [49]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2)

The **train_test_split** function from the sklearn.model_selection module is used to split a dataset into two subsets: one for training a machine learning model and one for testing its performance. Here's a breakdown of how it works and what the code you provided does:

##### Inputs:

* x: The input features of the dataset (e.g., independent variables).
* y: The target variable (e.g., dependent variable) corresponding to the input features.

###### Parameters:

* test_size=0.2: This parameter specifies the proportion of the dataset to include in the test split. Here, 0.2 means 20% of the data will be used for testing, and the remaining 80% will be used for training.

#####  Outputs:

* x_train: The training set features.
* x_test: The test set features.
* y_train: The training set target variable.
* y_test: The test set target variable.

##### Process:

The function randomly shuffles the dataset and splits it into two subsets according to the specified test_size.

In [50]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
from sklearn.preprocessing import OneHotEncoder

In [51]:
ohe=OneHotEncoder()
ohe.fit(x[['name','company','fuel_type']])

OneHotEncoder()

In [52]:
ohe.categories_

[array(['Audi A3 Cabriolet', 'Audi A4 1.8', 'Audi A4 2.0', 'Audi A6 2.0',
        'Audi A8', 'Audi Q3 2.0', 'Audi Q5 2.0', 'Audi Q7', 'BMW 3 Series',
        'BMW 5 Series', 'BMW 7 Series', 'BMW X1', 'BMW X1 sDrive20d',
        'BMW X1 xDrive20d', 'Chevrolet Beat', 'Chevrolet Beat Diesel',
        'Chevrolet Beat LS', 'Chevrolet Beat LT', 'Chevrolet Beat PS',
        'Chevrolet Cruze LTZ', 'Chevrolet Enjoy', 'Chevrolet Enjoy 1.4',
        'Chevrolet Sail 1.2', 'Chevrolet Sail UVA', 'Chevrolet Spark',
        'Chevrolet Spark 1.0', 'Chevrolet Spark LS', 'Chevrolet Spark LT',
        'Chevrolet Tavera LS', 'Chevrolet Tavera Neo', 'Datsun GO T',
        'Datsun Go Plus', 'Datsun Redi GO', 'Fiat Linea Emotion',
        'Fiat Petra ELX', 'Fiat Punto Emotion', 'Force Motors Force',
        'Force Motors One', 'Ford EcoSport', 'Ford EcoSport Ambiente',
        'Ford EcoSport Titanium', 'Ford EcoSport Trend',
        'Ford Endeavor 4x4', 'Ford Fiesta', 'Ford Fiesta SXi', 'Ford Figo',
        '

##### OneHotEncoder
* The OneHotEncoder is used to transform categorical data into a one-hot numeric array. This means it converts each category value into a new column and assigns a 1 or 0 (True/False) to the column.

* The OneHotEncoder from sklearn.preprocessing is used to convert categorical variables into a format that can be provided to machine learning algorithms to do a better job in prediction.

* Here’s a breakdown of the code you provided:

###### Import and Initialization:

**ohe=OneHotEncoder()**: This initializes an instance of the OneHotEncoder class.

###### Fitting:

**ohe.fit(x[['name','company','fuel_type']])**: This fits the OneHotEncoder to the specified columns in the dataset x, which are 'name', 'company', and 'fuel_type'.

## Creating a column transformer to transform categorical columns

In [53]:
from sklearn.compose import make_column_transformer
column_trans=make_column_transformer((OneHotEncoder(categories=ohe.categories_),['name','company','fuel_type']),
                                    remainder='passthrough')

###### OneHotEncoder:

**OneHotEncoder(categories=ohe.categories_)**: This uses the categories learned from the previous ohe.fit call to ensure consistent encoding.when TEST_data has some Categories that not in Train_data then model will through an error so to resolve it ,WE are passing ALL the categoriesto OneHotEncoder ,so that it know all the categories.


* ['name','company','fuel_type']: These are the columns to which the one-hot encoding will be applied.

###### Remainder:

remainder='passthrough': This indicates that all other columns not specified in the transformer should be passed through without any transformation.

#### Linear Regression Model

In [54]:
lr=LinearRegression()

### making pipeline

In [55]:
from sklearn.pipeline import make_pipeline
pipe=make_pipeline(column_trans,lr)

##### Pipeline:

**make_pipeline(column_trans, lr)**: This creates a pipeline that first applies the column transformer (column_trans) to the data and then fits the linear regression model (lr) to the transformed data.

In [56]:
pipe.fit(x_train,y_train)

C:\Users\ADMIN\AppData\Roaming\Python\Python311\site-packages\sklearn\compose\_column_transformer.py:1624: FutureWarning: 
The format of the columns of the 'remainder' transformer in ColumnTransformer.transformers_ will change in version 1.7 to match the format of the other transformers.
At the moment the remainder columns are stored as indices (of type int). With the same ColumnTransformer configuration, in the future they will be stored as column names (of type str).
To use the new behavior now and suppress this warning, use ColumnTransformer(force_int_remainder_cols=False).

  warnings.warn(


Pipeline(steps=[('columntransformer',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('onehotencoder',
                                                  OneHotEncoder(categories=[array(['Audi A3 Cabriolet', 'Audi A4 1.8', 'Audi A4 2.0', 'Audi A6 2.0',
       'Audi A8', 'Audi Q3 2.0', 'Audi Q5 2.0', 'Audi Q7', 'BMW 3 Series',
       'BMW 5 Series', 'BMW 7 Series', 'BMW X1', 'BMW X1 sDrive20d',
       'BMW X1 xDrive20d', 'Chevrolet Beat', 'Chevrolet Beat...
                                                                            array(['Audi', 'BMW', 'Chevrolet', 'Datsun', 'Fiat', 'Force', 'Ford',
       'Hindustan', 'Honda', 'Hyundai', 'Jaguar', 'Jeep', 'Land',
       'Mahindra', 'Maruti', 'Mercedes', 'Mini', 'Mitsubishi', 'Nissan',
       'Renault', 'Skoda', 'Tata', 'Toyota', 'Volkswagen', 'Volvo'],
      dtype=object),
                                                                            array(['Diesel', 'LPG', 'Petrol'], dtype=object)]),
                                                  ['name', 'company',
                                                   'fuel_type'])])),
                ('linearregression', LinearRegression())])

In [57]:
y_pred=pipe.predict(x_test)

In [58]:
y_pred

array([ 107290.06686668,  574380.15344162,  162875.32418974,
        440047.76716995,  589515.56709245,  350986.70982644,
       -222868.25527519,  775330.26595461,  367090.42449535,
        180443.53054962,  408761.13916324,  225974.21758334,
         40241.56266013,  282352.64257276,  290480.7291944 ,
        224603.36738126,  775330.26595461,  209099.19051485,
        355974.94008745,  576722.48283845,  288841.41324256,
        431058.18088791,  314402.76945198,  452800.0281083 ,
        232787.78723024,  274607.68688615,  -30478.85401224,
        234868.20923527,  113410.17996953,  670571.6975361 ,
        605159.1551784 ,  274440.40748075,  414552.76543705,
        718607.36099982,   63434.21376289,  790735.42817482,
        180443.53054962,  129367.77775346,  571722.50169124,
        406618.8086469 ,  610014.39034849,  131601.94989887,
        419556.4834605 ,   20306.41128978,  623323.42552026,
        197282.21602678,  608550.11307497,  374206.04208532,
        -70267.23816746,

### Checking R2 Score

In [59]:
r2_score(y_test,y_pred)

0.7202421801242185

### Finding the model with a **random_state** of TrainTestSplit where the model was found to give almost max as r2_score

In [60]:
score=[]
for i in range(1000):
    x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=i)
    lr=LinearRegression()
    pipe=make_pipeline(column_trans,lr)
    pipe.fit(x_train,y_train)
    y_pred=pipe.predict(x_test)
    score.append(r2_score(y_test,y_pred))

* The provided code is performing a robustness check by evaluating the performance of a linear regression model using cross-validation. Specifically, it repeatedly splits the dataset into training and testing sets, fits a linear regression model on the training set, and then evaluates it on the testing set, storing the R² scores for each iteration.
#### The random_state=i parameter in the train_test_split function ensures reproducibility by using a different seed for each iteration (from 0 to 999). This creates a different train-test split in each iteration, allowing the evaluation of the model's performance across various splits of the data.

In [61]:
import numpy as np

In [62]:
score

[0.65904340985128,
 0.4786077705982854,
 0.6207397118009343,
 0.5579870665329281,
 0.8188990179537847,
 0.7012191182904144,
 0.6276411147562568,
 0.6298988161339314,
 0.6720593531841353,
 0.5699706186481448,
 0.690335291187411,
 0.6086162217817972,
 0.4570484170741722,
 0.6674974737564106,
 0.6170174018837029,
 0.6437592782414592,
 0.487918737878307,
 0.665781686638292,
 0.6089308043982451,
 0.6399079124528706,
 0.6410698305676144,
 0.6211061369369246,
 0.7366487917219445,
 0.6403322118603012,
 0.5492186553108053,
 0.49869268006431333,
 0.6732783661733807,
 0.7035575081778327,
 0.7254702882278911,
 0.6010278527587157,
 0.6245205740233757,
 0.6330326182237591,
 0.7538194403080943,
 0.7854163896180785,
 0.5622422525927935,
 0.6813027297030931,
 0.5826069980673123,
 0.6170328901097296,
 0.5392911744896974,
 0.7164024663582518,
 0.3292142123193882,
 0.6733754296713984,
 0.573130456899225,
 0.6793103711902542,
 0.6604128295732372,
 0.6975060730887681,
 0.5258728734358504,
 0.58818260334462,

In [63]:
np.argmax(score)

433

In [64]:
score[np.argmax(score)]

0.8456957793726847

#### The best model is found at a certain random state 

In [65]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=np.argmax(score))
lr=LinearRegression()
pipe=make_pipeline(column_trans,lr)
pipe.fit(x_train,y_train)
y_pred=pipe.predict(x_test)
r2_score(y_test,y_pred)

0.8456957793726847

### Now make a predict with Random DATA

In [67]:
pipe.predict(pd.DataFrame([['Maruti Suzuki Swift','Maruti',2019,100,'Petrol']],columns=x_test.columns))

array([458934.66740154])

### Save the pipeline using pickle

In [68]:
import pickle

In [69]:
pickle.dump(pipe,open('LinearRegressionModel.pkl','wb'))

* The provided code serializes(saves) the trained pipeline object to a file using the pickle module. Here's a precise explanation:

**pickle.dump:**
This function serializes the object (in this case, pipe) and writes it to a file.
###### Parameters:
* **pipe:** The trained pipeline object that you want to save.
* **open('LinearRegressionModel.pkl', 'wb'):** Opens a file named LinearRegressionModel.pkl in write-binary mode ('wb'), preparing it to store the serialized object.